include/boost/url/impl/url.ipp

```diff
@@ -2046,22 +2047,92 @@ url&
 url::
 normalize()
 {
-#if 0
-    // scheme
-    {
-        char* it = s_
-            + offset(id_scheme);
-        auto last = it + offset(id_user);
-        if(it < last)
+    char* dest = s_;
+    char buf{'\0'};
+    int all_ids[] = {id_scheme, id_user, id_pass,
+                      id_host,  id_port, id_path,
+                      id_query, id_frag};
+    int next_id_offset = offset(id_scheme);
+    for (auto id: all_ids)
```

Comment:
I don't think this is right

Reply:
I'll go through that and compare to other functions to check how this can improve.

Reply:
Each of the 5 parts of the URL in the generic syntax uses its own charset for percent-encoding, to determine what is reserved. Scheme for example, has no escapes at all.

Reply:
The RFC says in [Section 6.2.2.2](https://datatracker.ietf.org/doc/html/rfc3986#section-6.2.2.2)

> In addition to the case normalization issue noted above, some URI producers percent-encode octets that do not require percent-encoding, resulting in URIs that are equivalent to their non-encoded counterparts. These URIs should be normalized by decoding any percent-encoded octet that corresponds to an unreserved character, as described in [Section 2.3](https://datatracker.ietf.org/doc/html/rfc3986#section-2.3).

I think there's a difference between necessary and sufficient conditions in these algorithms. Unlike the usual algorithm to produce URLs, normalization is not trying to validate or sanitize what needs to be encoded. Instead, it's trying to identify remove unnecessary encoding and explicitly defines these as the unreserved characters only.

Of course, we can expand that to other characters and decode whatever could be already decoded in the component. I'm fine with this because it would be much better in terms of avoiding false positives. It's just deviating slightly from the RFC though.

Reply:
Good news. I checked javascript's URL and it also decodes chars that are accepted in that component. So it's one more reason to do it. 

```js
new URL('http://%68examPle.com:80/path/../index.html?gliu%5B%5D=1');
```

```js
URL {
  href: 'http://hexample.com/index.html?gliu%5B%5D=1',
  origin: 'http://hexample.com',
  protocol: 'http:',
  username: '',
  password: '',
  host: 'hexample.com',
  hostname: 'hexample.com',
  port: '',
  pathname: '/index.html',
  search: '?gliu%5B%5D=1',
  searchParams: URLSearchParams { 'gliu[]' => '1' },
  hash: ''
}
```

This URL class is more focused on HTTP URLs though. For instance, port 80 is normalized to `""`.

Reply:
I guess once you break up the algorithm into 5 individual functions (scheme, authority, path, query, and fragment) then changing it will be easier. So we shouldn't worry too much about the character set.